File: ali/ebay/get_order_list.js
// vim: set ts=4 sw=4 :
//
// usage: casperjs get_order_list.js [--full] [--detailed]
var  utils = require('utils'),
    system = require('system'),
    fs = require('fs'),
    core = require('./core');
var casper = core.create_casper();
var delimiter= core.delimiter;

var orders =[],
    detailed_parse = false,
    full_parse = false;

var order_list="http://my.ebay.com/ws/eBayISAPI.dll?MyEbayBeta&CurrentPage=MyeBayNextSummary";

//casper.on('remote.message', function(msg) {
//        system.stderr.writeLine('remote message caught: ' + msg);
//});

if (casper.cli.get('full')){
  system.stderr.writeLine("crawling complete order list");
  full_parse=true;
}
if (casper.cli.get('detailed')){
  system.stderr.writeLine("crawling detailed list");
  detailed_parse=true;
}
if (casper.cli.get(0) == "help"){
  system.stderr.writeLine("usage: casperjs get_order_list.js [--full] [--detailed]");
  casper.exit()
}

//casper.on('remote.message', function(msg) {
//      this.echo('remote message caught: ' + msg);
//})
/*
 * selectors:
 * .querySelectorAll(".toolbar .ui-pager-label")[0].innerHTML.split(" ")[2] <- number of pages
 *
 *  document.querySelectorAll('.next') <- next button
 *    this.click('.next')
 *
 *
    function getOrders() {
    var links = document.querySelectorAll('.ae-order');
 */


// document.querySelector("div.filter").querySelectorAll("span.data-json")
function hasNextButton(){
    return document.querySelector('.next');
}
function clickNextButton(){
}

function getLastPage(){
    return document.querySelector('.pagination span').textContent.split(' ').pop()
}
function getCurrentPage(){
    return document.querySelector('.pagination span').textContent.split(' ')[1]
}

function getOptionSelection(){
    ret = [];
    document.querySelector('.arrow-down').click();
    [].forEach.call(document.querySelectorAll("[id*=OptionMenu] [role=menuitem]"),function (item){ret.push(item.textContent)})
    return ret;
}
function getCurrentOption(){
    return document.querySelector(".filter-content").textContent.trim()

}
// we always start with the current option

function getOptionLength(){
    // first we have to click the down arrow, then we can get the items
    document.querySelector('.arrow-down').click()
    // we click the first item, then the second, and so on
    return document.querySelectorAll("[id*=OptionMenu] [role=menuitem]").length
}

function atLastOption(current){
    function getOptionLength(){
        // first we have to click the down arrow, then we can get the items
        // we click the first item, then the second, and so on
        return document.querySelectorAll("[id*=OptionMenu] [role=menuitem]").length
    }
    return getOptionLength() == current ;
}

function atLastPage(){
    function getLastPage(){
        return document.querySelector('.pagination span').textContent.split(' ').pop()
    }
    function getCurrentPage(){
        return document.querySelector('.pagination span').textContent.split(' ')[1]
    }
    return getLastPage() == getCurrentPage();

}

function prepareOptions(){
    document.querySelector('.arrow-down').click()
}
function clickNextOption(current){
    //because nth-child counts from 1
    casper.click("[id*=OptionMenu] li:nth-child("+(1+current)+") [role=menuitem]")

    return 1+current
}

function getLinks(delimiter) {
    //console.log("in getLinks")
    var orders = [];
    var sel_orders = document.querySelectorAll('.paction');
    for (var i = 0, len = sel_orders.length; i < len; i++){
        sel_order = sel_orders[i]
        order_opts = sel_order.href.split('?')[1].split('&')
        var transid=0,
            itemid=0;
        for (var j = 0, optlen = order_opts.length; j < optlen; j++){
            cur_opt = order_opts[j].split('=');
            var k = cur_opt[0]
            var v = cur_opt[1]
            //console.log(k+" --> " + v)
            if (k == 'transId' || k == 'transactID' || k == 'transactionId'){
                transid=v;
            } else if (k == 'itemid' || k == 'iid' || k == 'item' || k == 'itemId' ) {
                itemid=v;
            }
        }
        orders.push(""+transid+delimiter+itemid)
        //console.log(""+transid+delimiter+itemid)
    }
    console.log(JSON.stringify(orders))
    return orders

}
function onlyUnique(value, index, self) { 
        return self.indexOf(value) === index;
}

var current_option=0;
system.stderr.writeLine("loading cookie file")
function parse_orders(){
    system.stderr.writeLine("waiting for selector .item-list-all");
    //casper.waitForSelector('.item-list-all',function (){
        system.stderr.writeLine("waiting for timeout");
        casper.wait(1000,function (){

            system.stderr.writeLine("!!!!!!" + this.evaluate(getCurrentOption) + " Page "+ this.evaluate(getCurrentPage) + " of " + this.evaluate(getLastPage) )
            Array.prototype.push.apply(orders,this.evaluate(getLinks,delimiter));
            // go through all pages for all available options
            if (full_parse){
                this.click('.arrow-down')
                this.waitForSelector("[id*=OptionMenu] [role=menuitem]",function (){
                //casper.wait(500,function(){
                    if (this.evaluate(atLastPage)){
                        system.stderr.writeLine("at last page of current Option");
                        if (this.evaluate(atLastOption,current_option)){
                            system.stderr.writeLine("at last Option of Last page");
                        }else{

                                system.stderr.writeLine("current option " + current_option);
                                current_option = clickNextOption(current_option)
                                casper.capture("click-next.png")
                                system.stderr.writeLine("clicked next option ( at "+current_option+" of " +this.evaluate(getOptionLength)+")");
                                casper.wait(3000,parse_orders)
                         //   })
                        }
                        casper.echo(JSON.stringify(orders.filter(onlyUnique)));
                    }else{
                        casper.click('.next')
                        system.stderr.writeLine("clicked next")
                        casper.wait(3000,parse_orders)
                    }
                });
            }else{
                casper.echo(JSON.stringify(orders.filter(onlyUnique)));
            }
        });
    //},core.open_failed,5000);
}

phantom.cookies = core.load_cookies();
casper.start().thenOpen(order_list,parse_orders)
    //orders = (this.evaluate(getLinks));
    //system.stderr.writeLine(JSON.stringify(orders));
casper.run()

File: ali/ali/get_order_list.py
#!/usr/bin/env python
# vim: set ts=4 sw=4 
# TODO: this does not yet work
from core import load_config,save_db,load_db
from splinter import Browser
import time
from pprint import pprint
#order_list="http://trade.aliexpress.com/orderList.htm"
order_list='http://trade.alibaba.com/order_list.htm'
ali_website="http://aliexpress.com/"
customHeaders={"Referer":ali_website,"DNT":"1"}

config = load_config()
aconf=config['ali']

with Browser("firefox",user_agent=config['user_agent']) as browser:
    browser.visit(ali_website)
    browser.cookies.add(load_db(aconf['cookie_file']))
    pprint(browser.cookies.all())
    browser.find_by_css('.nav-user-account').mouse_over()
    time.sleep(10)
    browser.click_link_by_href(order_list)
    pprint(browser.cookies.all())
    time.sleep(20)




"""
// partly from https://code.google.com/p/phantomjs/issues/attachmentText?id=716&aid=7160000000&name=alipromo.js&token=78bfXm_YEKbtdwqWmkdPRGO7_fU%3A1352647368420
     
var casper = require('casper').create(),
    utils = require('utils'),
    system = require('system'),
    fs = require('fs');
var ali_home= system.env.HOME + "/.ali";
var orders =[],
    full_parse = true,
    cookie_file = ali_home+'/ali.cookies';
var order_list="http://trade.aliexpress.com/orderList.htm";
var ali_website="http://aliexpress.com/";
customHeaders={"Referer":ali_website,"DNT":"1"};
casper.options.stepTimeout = 10000;
casper.userAgent("Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/37.0.2049.0 Safari/537.36");

casper.on('page.error', function(err) {
      this.echo(JSON.dump({"error":"Complete callback has failed: " + err}));
      system.exit(1);
});

if (casper.cli.args.length === 0){
  system.stderr.writeLine("crawling first page")
  full_parse=false
}else{
  system.stderr.writeLine("crawling complete order list")
  full_parse=true
}
//casper.on('remote.message', function(msg) {
//      this.echo('remote message caught: ' + msg);
//})
/*
 * selectors:
 * .querySelectorAll(".toolbar .ui-pager-label")[0].innerHTML.split(" ")[2] <- number of pages
 *
 *  document.querySelectorAll('.next') <- next button
 *    this.click('.next')
 *
 *
    function getOrders() {
    var links = document.querySelectorAll('.ae-order');
 */

function hasNextButton(){
    return document.querySelector('.next');
}
function getPageText(){
    return document.querySelector('.pages-total').textContent;
}
function getLinks() {
    var orders = [];
    var sel_orders = document.querySelectorAll('.ae-order');
    for (var i = 0, len = sel_orders.length; i < len; i++){
        sel_order = sel_orders[i]
        order = parseInt(sel_order.getAttribute('data-orderid'),10)
        orders.push(order)
    }
    //.other-info .sell-sp-main (what guarantees give the seller)
    return orders

}
function open_failed(){
    this.echo(JSON.stringify({"error":"not logged in!"})).exit(1)
}

system.stderr.writeLine("loading cookie file")
function parse_orders(){
    system.stderr.writeLine("waiting for selector");
    casper.waitForSelector('.ae-order',function(){
        system.stderr.writeLine(this.evaluate(getPageText))
        Array.prototype.push.apply(orders,this.evaluate(getLinks));
        if (full_parse ){
            if (! this.evaluate(hasNextButton)){
                system.stderr.writeLine("at last page!")
                casper.echo(JSON.stringify(orders));
                casper.exit()
            }
            this.click('.next');
            system.stderr.writeLine("clicked next")
            casper.wait(3000,parse_orders)
        }else{
            casper.echo(JSON.stringify(orders));
        }
    },open_failed,4000);
}
try{
    var data = fs.read(cookie_file)
    phantom.cookies = JSON.parse(data)
    casper.start(order_list,{headers:customHeaders})
    parse_orders()
        //orders = (this.evaluate(getLinks));
        //system.stderr.writeLine(JSON.stringify(orders));
    casper.run()
}catch(e){
    this.echo(JSON.stringify({"error":"cannot load cookies"})).exit(1)
}
"""
